### PR TITLE
Feature: Add Oracle Pricing Data

### PR DESCRIPTION
**Pull request by jfarid27**
# Description

- Adds on chain oracle pricing data. The on chain oracle should properly compute prices from whitelisted tokens only and return zero for non-whitelisted tokens.
- The code should only hit the RPC once every hour to update the tokens.

# QA/Testing

We are getting good data locally with my personal RPCs, but I would like to see what we return from the deployment RPC as a full test.

Screenshot from 2024-10-09 19-33-04

### CODE REVIEW

File: src/EventHandlers/NFPM.ts [hunk 1]
```diff
@@ -30,4 +33,23 @@ NFPM.Transfer.handler(async ({ event, context }) => {
   };
 
   context.NFPM_Transfer.set(entity);
+
+  const lastUpdated = getPricesLastUpdated(event.chainId);
```
Comment (jfarid27) [inline src/EventHandlers/NFPM.ts:37]: This section will check prices on new position mint.

File: src/EventHandlers/Pool.ts [hunk 2]
```diff
@@ -291,21 +285,36 @@ Pool.Sync.handlerWithLoader({
...
+      const blockDatetime = new Date(event.block.timestamp * 1000);
+      const timeDelta = CHAIN_CONSTANTS[event.chainId].oracle.updateDelta * 1000;
+
+      const tokensNeedUpdate = !lastUpdated || (blockDatetime.getTime() - lastUpdated.getTime()) > timeDelta;
+
+      const oracleAvailable = CHAIN_CONSTANTS[event.chainId].oracle.startBlock < event.block.number;
+
+      if (tokensNeedUpdate && oracleAvailable) {
+        try {
+          await set_whitelisted_prices(event.chainId, event.block.number, blockDatetime, context);
```
Comment (jfarid27) [inline src/EventHandlers/Pool.ts:301]: This section checks prices on new pool sync.

File: src/EventHandlers/NFPM.ts [hunk 3]
```diff
@@ -30,4 +33,23 @@ NFPM.Transfer.handler(async ({ event, context }) => {
...
+  // Check if the oracle is available and deployed
+  const oracleAvailable = CHAIN_CONSTANTS[event.chainId].oracle.startBlock < event.block.number;
+  if (!oracleAvailable) return;
+
+  // Only update the prices if the last update was more than updateDelta seconds ago.
+  const timeDelta = CHAIN_CONSTANTS[event.chainId].oracle.updateDelta * 1000;
+
+  if (!lastUpdated || (blockDatetime.getTime() - lastUpdated.getTime()) > timeDelta) {
+    try {
+      await set_whitelisted_prices(event.chainId, event.block.number, blockDatetime, context);
```
Comment (jfarid27) [inline src/EventHandlers/NFPM.ts:49]: This section checks prices on new position mint.

File: src/constants/baseWhitelistedTokens.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "address": "0x4200000000000000000000000000000000000006",
+    "symbol": "WETH",
+    "unit": "ether",
+    "createdBlock": 0
+  },
+  {
+    "address": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
+    "symbol": "USDC",
+    "unit": "ether",
+    "createdBlock": 2797221
```
Comment (moose-code) [inline src/constants/baseWhitelistedTokens.json:36]: Very few whitelisted tokens so far, I guess we will increase the size of this list?

File: src/PriceOracle.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+/**
+ * Reads the prices of specified tokens from a price oracle contract.
+ *
+ * This function interacts with a blockchain price oracle to fetch the current
+ * prices of a list of token addresses. It returns them as an array of strings.
+ * 
+ * @note: See https://github.com/ethzoomer/optimism-prices for underlying smart contract
+ * implementation.
```
Comment (moose-code) [inline src/PriceOracle.ts:27]: Nice code!! This file looks good

File: src/PriceOracle.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    prices.forEach((price, index) => {
+        pricesByAddress.set(addresses[index], Number(utils.fromWei(price, units[index])));
+    });
+    
+    for (const token of tokenData) {
+        const price = pricesByAddress.get(token.address) || 0;
+        
+        // Get or create Token entity
+        let tokenEntity = await context.Token.get(token.address);
```
Comment (moose-code) [inline src/PriceOracle.ts:102]: Heads up reading from db will decrease speed slightly but this is only really here and there so should be fine.

File: schema.graphql [hunk 7]
```diff
@@ -404,3 +405,14 @@ type CLPool_Swap {
   timestamp: Timestamp! @index
   chainId: Int! @index
 }
+
+scalar Decimal
+
+type TokenPrice {
+  id: ID! # Unique identifier for the record, could be a combination of token address and chain ID
```
Comment (moose-code) [inline schema.graphql:412]: chain id , token address and block number it seems. Maybe block timestamp is nicer than block number in future?

File: config.yaml [hunk 8]
```diff
@@ -57,10 +57,6 @@ contracts:
     handler: src/EventHandlers/Gauge.ts
     events:
       - event: NotifyReward(address indexed from, uint256 amount)
-  - name: PriceFetcher
-    handler: src/EventHandlers/PriceFetcher.ts
-    events:
-      - event: PriceFetched(address indexed token, uint256 price)
```
Comment (moose-code) [inline config.yaml:63]: Happy to get rid of these. There is more breadth these are providing, but I guess we'll increase the whitelist list size

**Review by moose-code — APPROVED** [review]
Nice stuff, running locally my side and also looking good. I want to bump the version then I will deploy

**Pull request merged**